django/middleware/csrf.py

```diff
@@ -196,6 +197,37 @@ def _set_token(self, request, response):
             # Set the Vary header since content varies with the CSRF cookie.
             patch_vary_headers(response, ('Cookie',))
 
+    def _origin_verified(self, request):
+        good_origin = '%s://%s' % (
+            'https' if request.is_secure() else 'http',
+            request.get_host(),
+        )
+        allowed_origins = [good_origin]
+        # Tuples of (allowed scheme, allowed netloc), where all
+        # subdomains of netloc are allowed.
+        allowed_origin_subdomains = []
+        for trusted_origin in settings.CSRF_TRUSTED_ORIGINS:
+            if '*' in trusted_origin:
+                parsed = urlparse(trusted_origin)
+                allowed_origin_subdomains.append(
+                    (parsed.scheme, parsed.netloc.lstrip('*'))
+                )
+            else:
+                allowed_origins.append(trusted_origin)
+        # Reject the request if the origin header doesn't match an
+        # allowed value.
+        expected_origin = request.META['HTTP_ORIGIN']
```

Comment:
Small nitpick, but can we please rename this variable? "expected" to me in this context means that the server would expect this value. I'd like a variable naming here that indicates it is coming from the client. The referer code below uses `referer` or `request_*`. I guess `origin` as well as `request_origin` would be fine (not sure how to call `expected_scheme` & `expected_netloc` though). Open to naming ideas :)